field_audit/models.py

```diff
@@ -0,0 +1,350 @@
+        return self.filter(  # pragma: no cover
+            changed_by__contains={"user_type": user_type, "username": username},
+        )
+
+    # "no branch" note: tests only run on postgres _or_ sqlite, never both
+    if check_engine_sqlite():  # pragma: no branch
+        _by_type_and_username = by_type_and_username
+
+        def by_type_and_username(self, user_type, username):
+            """Support SQLite (for development) and Oracle (because it comes
+            along for free).
+
+            If these DB flavors are not needed at import time, this "extra db"
+            support is never defined.
+            """
+            if check_engine_sqlite(settings.DATABASES[self.db]["ENGINE"]):
+                # Oracle and SQLite do not support `contains` queries
+                # see: https://docs.djangoproject.com/en/4.0/topics/db/queries/#std:fieldlookup-jsonfield.contains  # noqa: E501
+                return self.filter(
+                    changed_by__user_type=user_type,
+                    changed_by__username=username,
+                )
+            # "no cover" note: tests only run on postgres _or_ sqlite, never
+            # both
+            return self._by_type_and_username(user_type, username)  # pragma: no cover  # noqa: E501
+
+
+class DefaultAuditEventManager(AuditEventManager):
+    """Default Manager for the AuditEvent model. Contains convenience methods
+    for the default auditors, which may not be desirable to subclass if
+    downstream projects wish to define custom auditor chains.
+    """
+
+    def by_system_user(self, username):
+        system_types = [USER_TYPE_TTY, USER_TYPE_PROCESS]
+        return self.filter(
+            changed_by__user_type__in=system_types,
+            changed_by__username=username,
+        )
+
+    def by_tty_user(self, username):
+        return self.by_type_and_username(USER_TYPE_TTY, username)
+
+    def by_process_user(self, username):
+        return self.by_type_and_username(USER_TYPE_PROCESS, username)
+
+    def by_request_user(self, username):
+        return self.by_type_and_username(USER_TYPE_REQUEST, username)
+
+
+def get_manager(attr_suffix, default):
+    """Returns an instantiated manager, possibly one defined in settings.
+
+    If the manager is defined in settings, it must be a subclass of
+    ``django.db.models.Manager``.
+
+    :param attr_suffix: Suffix (appended to ``FIELD_AUDIT_``) of settings
+        attribute for the manager class path.
+    :parma default: Manager class. Returned if attribute isn't defined.
+    """
+    settings_attr = f"FIELD_AUDIT_{attr_suffix}"
+    try:
+        class_path = getattr(settings, settings_attr)
+    except AttributeError:
+        return default()
+    desc = f"{settings_attr!r} value"
+    return class_import_helper(class_path, desc, models.Manager)()
+
+
+def get_date():
+    return datetime.utcnow()
+
+
+class AuditEvent(models.Model):
+    # id = models.BigAutoField(primary_key=True)
```

Comment:
Is this intentional?

Reply:
Yes, this is an artifact of me not making up my mind on the whole `AutoField` thing. Would be an interesting discussion to have, I think.

Reply:
> Would be an interesting discussion to have, I think
Curious what this discussion is?

Reply:
Prior to Django 3.2, if you didn't explicitly configure a `primary_key` field, Django would implicitly add one for you using `AutoField`. Since 3.2, if you don't specify a primary key column (i.e. you use the "implicit behavior" _and_ you don't specify either `YourAppConfig.default_auto_field` or `settings.DEFAULT_AUTO_FIELD`), Django will emit warnings when you call `makemigrations`.  I didn't want this app to **hard define** which AutoField a downstream project must use (`AutoField` vs. `BigAutoField`), so I added a (still undocumented) app setting `FIELD_AUDIT_AUTO_FIELD` to provide downstream projects the ability to choose. I'm still undecided on the merits of this. On one hand it might be overkill (just hard-require `BigAutoField`), on the other hand since Django doesn't force this decision, it seems reasonable to leave it up to downstream "consumers".

Reply:
For a library like this I think it makes sense to use `BigAutoField` rather than leaving it to the consumer. It makes sense for Django to leave it for the consumer since Django doesn't know what the model will do or how many records to expect etc. Since we DO know that for this model there is less need to leave the decision open.

Reply:
Thanks for that feedback @snopoke, I'm going to rip this part out and just explicitly use `BigAutoField`.

Reply:
Changed in f122828.